README.md
# [Forum Icons](http://www.chack1172.altervista.org/Projects/MyBB-18/Forum-Icons.html)
Category: [MyBB 1.8](http://www.chack1172.altervista.org/Projects/MyBB-18/)
Author: [chack1172](http://en.chack1172.altervista.org)

Forum Icons is a plugin that enhances your MyBB forum experience by allowing you to add unique icons to each forum.

## Installation
1. Upload contents of the folder 'src' to your MyBB board directory
2. Access your Admin Control Panel (ACP).
3. Navigate to `Plugins` and select `Install & Activate` on `Forum Icons`.

## Managemant
Managing forum icons is easy. Follow these steps:
1. Access your Admin Control Panel (ACP).
2. Go to `Forum & Posts` and select `Forum Icons`.
3. Insert the URL of the icon to each forum.

## Changelog

Check out the [CHANGELOG](CHANGELOG.md) to see the history of updates and new features.

## License
This project is licensed under the MIT License. Please see [License File](LICENSE) for more information.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## [5.1] - 2023-06-08

### Added
- Added changelog file
- Added template to templatelist to load

### Changed
- Use MyBB datacache to store icons instead of a static variable.
- Removed check for forum_icons table existance on admin page
- Reduced queries on ACP forum icons list page

### Fixed
- PHP warnings - Thanks to @SvePu


src/admin/modules/forum/icons.php
<?php

if(!defined("IN_MYBB"))
	die("Direct initialization of this file is not allowed.");

$lang->load("forum_icons");

if(isset($_POST['save_images']) && !empty($mybb->input['image']) && is_array($mybb->input['image'])) {
    foreach($mybb->input['image'] as $fid => $image) {
        $image = htmlspecialchars($db->escape_string($image));
        if($db->num_rows($db->simple_select("forum_icons", "*", "fid='{$fid}'")) > 0)
            $db->update_query("forum_icons", array("image" => $image), "fid={$fid}");
        elseif(!empty($image))
            $db->insert_query("forum_icons", array("fid" => $fid, "image" => $image));
    }

    ficons_update_cache();

    log_admin_action($lang->ficons_log);

    flash_message($lang->ficons_saved, 'success');
    admin_redirect("index.php?module=forum-icons");
}

$page->add_breadcrumb_item($lang->ficons_title, "index.php?module=forum-icons");
$page->extra_header .= <<<EOT
<style>
.preview {
    max-width: 200px;
    max-height: 150px;
}
</style>
<script type="text/javascript">
    $(document).ready(function() {
        $("form#images_url").on("keyup", "input.imageurl", function () {
            var content = $(this).val();
            $(this).parent("td").parent("tr").find(".preview").attr("src", content);
        });
    });
</script>
EOT;
$page->output_header($lang->ficons_title);

$form = new Form("index.php?module=forum-icons", "post", "images_url");
$form_container = new FormContainer($lang->ficons_manage);

$form_container->output_row_header($lang->ficons_forum, array("style" => "width: 50%"));
$form_container->output_row_header($lang->ficons_image, array("style" => "width: 25%", "class" => "align_center"));
$form_container->output_row_header($lang->ficons_preview, array("style" => "width: 25%", "class" => "align_center"));



$icons = [];
$query = $db->simple_select("forum_icons", "*");
while ($icon = $db->fetch_array($query)) {
	$icons[$icon['fid']] = $icon;
}

build_forums_list($form_container);

$submit_options = array();

if($form_container->num_rows() == 0){
	$form_container->output_cell($lang->ficons_no_forum, array('colspan' => 3));
	$form_container->construct_row();
	$submit_options = array('disabled' => true);
}

$form_container->end();

$buttons[] = $form->generate_submit_button($lang->ficons_save, array("name" => "save_images"));
$buttons[] = $form->generate_reset_button($lang->ficons_reset);

$form->output_submit_wrapper($buttons);

$form->end();

$page->output_footer();

function build_forums_list($form_container, $pid=0, $depth=1) {
    global $mybb, $lang, $db, $sub_forums, $form, $icons;
	static $forums_by_parent;

    if(!is_array($forums_by_parent)) {
		$forum_cache = cache_forums();

		foreach($forum_cache as $forum)
			$forums_by_parent[$forum['pid']][$forum['disporder']][$forum['fid']] = $forum;
	}

	if(!is_array($forums_by_parent[$pid]))
		return;

	foreach($forums_by_parent[$pid] as $children) {
		foreach($children as $forum) {
			if (isset($icons[$forum['fid']])) {
				$image = $icons[$forum['fid']];
			} else {
				$image = ['image' => ''];
			}

			$forum['name'] = preg_replace("#&(?!\#[0-9]+;)#si", "&amp;", $forum['name']);

			if($forum['type'] == "c" && $depth == 1) {
				$form_container->output_cell("<div style=\"padding-left: ".(40*($depth-1))."px;\"><strong>{$forum['name']}</strong></div>", array("colspan" => "3", "style" => "height:33px"));
                $form_container->construct_row();

				if($forums_by_parent[$forum['fid']])
					build_forums_list($form_container, $forum['fid'], $depth+1);
			}
			else {
                $form_container->output_cell("<div style=\"padding-left: ".(40*($depth-1))."px;\">{$forum['name']}</div>");
                $form_container->output_cell($form->generate_text_box("image[{$forum['fid']}]", $image['image'], array("class" => "imageurl")), array("class" => "align_center"));
                $form_container->output_cell("<img class=\"preview\" src=\"{$image['image']}\" alt=\"{$lang->ficons_no_preview}\">", array("class" => "align_center"));
                $form_container->construct_row();

                if(isset($forums_by_parent[$forum['fid']]))
					build_forums_list($form_container, $forum['fid'], $depth+1);
			}
		}
	}
}


src/inc/plugins/ficons.php
<?php

if(!defined('IN_MYBB')) {
	die('This file cannot be accessed directly.');
}

if(defined('IN_ADMINCP')) {
    $plugins->add_hook("admin_forum_menu", "ficons_sub_menu");
    $plugins->add_hook("admin_forum_action_handler", "ficons_action");
} else {
	$plugins->add_hook("global_start", "ficons_templates");
    $plugins->add_hook("build_forumbits_forum", "ficons_show");
}

function ficons_info() {
    return array(
        "name" => "Forum Icons",
        "description" => "Add unique icons to each forum",
        "website" => "",
        "author" => "chack1172",
        "authorsite" => "http://chack1172.altervista.org/",
        "version" => "5.1",
        "compatibility" => "18*",
        "codename" => "ficons"
    );
}

function ficons_install() {
    global $db;

    $collation = $db->build_create_table_collation();

	if(!$db->table_exists("forum_icons")) {
		$db->write_query("CREATE TABLE `".TABLE_PREFIX."forum_icons` (
			`id` int(10) UNSIGNED NOT NULL auto_increment,
			`fid` int(10) UNSIGNED NOT NULL,
            `image` varchar(500) NOT NULL default '',
			PRIMARY KEY  (`id`)
		) ENGINE=MyISAM{$collation}");
	}

    $template = '<img src="{$forum_icon}" alt="" style="float: left; max-width: 200px; max-height: 150px;padding-right: 10px">';

    $insert_array = array(
        'title' => 'forum_icons',
        'template' => $db->escape_string($template),
        'sid' => '-1',
        'version' => '',
        'dateline' => time()
    );

    $db->insert_query('templates', $insert_array);

    $setting_group = array(
        'name' => 'ficons',
        'title' => 'Forum Icons Settings',
        'description' => 'Here you can change settings of plugin Forum Icons',
        'disporder' => 5,
        'isdefault' => 0
    );

    $gid = $db->insert_query("settinggroups", $setting_group);

    $setting_array = array(
        'ficons_visible' => array(
            'title' => 'Show icons?',
            'description' => 'Would you show icons in forum list?',
            'optionscode' => 'yesno',
            'value' => 1,
            'disporder' => 1
        ),
    );


    foreach($setting_array as $name => $setting) {
        $setting['name'] = $name;
        $setting['gid'] = $gid;

        $db->insert_query('settings', $setting);
    }

    rebuild_settings();

	// Create datacache entry
	ficons_update_cache([]);
}

function ficons_is_installed() {
    global $db;

    return $db->table_exists("forum_icons");
}

function ficons_uninstall() {
    global $db, $cache;

    $db->drop_table("forum_icons");
    $db->delete_query("templates", "title = 'forum_icons'");

    $db->delete_query('settings', "name = 'ficons_visible'");
    $db->delete_query('settinggroups', "name = 'ficons'");

    rebuild_settings();

    $cache->delete('forum_icons');
}

function ficons_activate() {
    global $mybb;

    require_once MYBB_ROOT."/inc/adminfunctions_templates.php";

    find_replace_templatesets(
        "forumbit_depth2_forum",
        "#" . preg_quote('{$forum[\'name\']}') . "#i",
        '{$forum[\'icon\']}{$forum[\'name\']}'
    );

    find_replace_templatesets(
        "forumbit_depth2_cat",
        "#" . preg_quote('{$forum[\'name\']}') . "#i",
        '{$forum[\'icon\']}{$forum[\'name\']}'
    );
}

function ficons_deactivate() {
    require_once MYBB_ROOT."/inc/adminfunctions_templates.php";

    find_replace_templatesets(
        "forumbit_depth2_forum",
        "#" . preg_quote('{$forum[\'icon\']}') . "#i",
        ''
    );

    find_replace_templatesets(
        "forumbit_depth2_cat",
        "#" . preg_quote('{$forum[\'icon\']}') . "#i",
        ''
    );
}

function ficons_sub_menu(&$sub_menu) {
    global $lang;

    $lang->load("forum_icons");

    $sub_menu[] = array("id" => "icons", "title" => $lang->ficons_title, "link" => "index.php?module=forum-icons");
}

function ficons_action(&$actions) {
    $actions['icons'] = array('active' => 'icons', 'file' => 'icons.php');
}

function ficons_show(&$forum) {
    global $templates, $forum_url, $mybb;

	$forum['icon'] = '';
    if($mybb->settings['ficons_visible'] == 1) {
        $forum_icon = ficons_get_icon($forum['fid']);
        if(!empty($forum_icon)) {
            eval("\$forum['icon'] = \"".$templates->get("forum_icons")."\";");
        }
    }
}

function ficons_get_icon($fid) {
    global $cache;
	$iconsCache = $cache->read('forum_icons');
    if (!is_array($iconsCache)) {
        $iconsCache = ficons_update_cache();
    }
    if (isset($iconsCache[$fid])) {
        return $iconsCache[$fid];
    } else {
        return '';
    }
}

function ficons_update_cache($values = null)
{
	global $cache, $db;
	if ($values === null) {
		$query = $db->simple_select('forum_icons', 'fid, image');
		$values = [];
		while ($ficon = $db->fetch_array($query)) {
            $values[$ficon['fid']] = $ficon['image'];
        }
	}
	$cache->update('forum_icons', $values);

	return $values;
}

function ficons_templates()
{
	global $templatelist;

	if (defined('THIS_SCRIPT') && in_array(THIS_SCRIPT, ['index.php', 'forumdisplay.php'])) {
		if(isset($templatelist)) {
			$templatelist .= ',';
		} else {
			$templatelist = '';
		}
		$templatelist .= 'forum_icons';
	}
}

function reload_forum_icons()
{
	ficons_update_cache();
}
